cas-server-core/src/main/java/org/jasig/cas/authentication/support/DefaultCasAttributeEncoder.java

```diff
@@ -0,0 +1,226 @@
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL);
+    }
+
+    /**
+     * Encode and encrypt pgt using the public key
+     * supplied by the service. The result is base64 encoded
+     * and put into the attributes collection again, overwriting
+     * the previous value.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     */
+    protected void encodeAndEncryptProxyGrantingTicket(final Map<String, Object> attributes,
+                                                       final Map<String, String> cachedAttributesToEncode) {
+        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
+                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);
+    }
+
+    /**
+     * Encrypt, encode and put the attribute into attributes map.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param cachedAttributeName the cached attribute name
+     */
+    protected void encryptAndEncodeAndPutIntoAttributesMap(final Map<String, Object> attributes,
+                                                           final Map<String, String> cachedAttributesToEncode,
+                                                           final String cachedAttributeName) {
+        final String cachedAttribute = cachedAttributesToEncode.remove(cachedAttributeName);
+        if (StringUtils.isNotBlank(cachedAttribute)) {
+            logger.debug("Retrieved [{}] as a cached model attribute...", cachedAttributeName);
+            final String encodedValue = CompressionUtils.encryptAndEncodeBase64(cachedAttribute, this.cipher);
+            attributes.put(cachedAttributeName, encodedValue);
+            logger.debug("Encrypted and encoded [{}] as an attribute to [{}].", cachedAttributeName, encodedValue);
+        } else {
+            logger.debug("[{}] is not available as a cached model attribute to encrypt...", cachedAttributeName);
+        }
+    }
+
+    /**
+     * Initialize the cipher with the public key
+     * and then start to encrypt select attributes.
+     *
+     * @param attributes the attributes
+     * @param cachedAttributesToEncode the cached attributes to encode
+     * @param service the service
+     */
+    protected void encodeAttributesInternal(final Map<String, Object> attributes,
+                                            final Map<String, String> cachedAttributesToEncode,
+                                            final RegisteredService service) {
+        try {
+            if (!initializeCipherBasedOnServicePublicKey(service)) {
+                return;
+            }
+            encodeAndEncryptCredentialPassword(attributes, cachedAttributesToEncode);
```

Comment:
No, you're right. But if you are going down the route of overriding things, you can always code your own encoder and do everything in plain text. The objective here, as I understand, is not to lock down the configuration 100%. While the code here perfectly encrypts and encodes all sensitive data, I'd imagine the overriding ability to become useful later on, during troubleshooting and dealing with encryption issues, and/or to accommodate future use cases, such that one could *only* override the piece that is causing trouble or is in question. For that reason, I am not entirely sold on the finality thing here. [Otherwise, we could mark the entire class as final and go about our own way]

The other issue, that is closely related to the finality of the class is the method separation. If I didn't have the dedicated methods for instance, I'd have to have something like this as is suggested:

```java
protected void encodeAttributesInternal(final Map<String, Object> attributes,
                                            final Map<String, String> cachedAttributesToEncode,
                                            final RegisteredService service) {
        try {
            if (!initializeCipherBasedOnServicePublicKey(service)) {
                return;
            }
            encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);

            encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
                CasViewConstants.MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL);
        } catch (final Exception e) {
            throw new RuntimeException(e);
        }
    }
```

Two issues:

1. It's simply not as clear. Would you rather read:

```java
encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
                CasViewConstants.MODEL_ATTRIBUTE_NAME_PRINCIPAL_CREDENTIAL);
```

or would you rather read:

```java
encodeAndEncryptCredentialPassword(attributes, cachedAttributesToEncode);
```

2. Lets say I wanted to override the encoder to do "something" with the PGT encryption. If I have dedicated methods, all I have to do is to override:

```java
protected void encodeAndEncryptProxyGrantingTicket(final Map<String, Object> attributes,
                                                       final Map<String, String> cachedAttributesToEncode) {
        encryptAndEncodeAndPutIntoAttributesMap(attributes, cachedAttributesToEncode,
                CasViewConstants.MODEL_ATTRIBUTE_NAME_PROXY_GRANTING_TICKET);
    }
```

If I dont have that dedicated method, I'd have to override the entire `encodeAttributesInternal` which is not something I wanted. I just want that one piece.

Makes sense?